_posts/2016-09-14-programming-model.md

```diff
@@ -0,0 +1,128 @@
+
+A pipeline state objects (PSO) encapsulates all the dependent states of your shader program (with the program itself). You declare in advance what kind of vertex attributes will come as input, what render targets to draw to, with what blending modes, what textures/samplers to use, etc. This allows the driver to compile the most efficient code for the set of states. Batching becomes trivial - just send multiple draw calls with the same PSO to get the maximum performance. The downside here is the inconvenience and restrictive nature of these forward state declarations.
+
+Our PSOs are defined with macros like this:
+{% highlight Rust %}
+gfx_defines!{
+    vertex Vertex { // Defininig a vertex buffer format. You can have multiple vertex buffers having different attributes in them.
+        pos: [f32; 2] = "a_Pos",
+        color: [f32; 3] = "a_Color",
+    }
+
+    pipeline pipe {
+        vbuf: gfx::VertexBuffer<Vertex> = (), // our vertex buffer
+        out: gfx::RenderTarget<ColorFormat> = "Target0", // the output target, no blending involved
+    }
+}
+{% endhighlight %}
+
+When making a draw call, you provide the generated `Data` struct instance pointing to the used resources and views:
+{% highlight Rust %}
+encoder.draw(&slice, &pso, &pipe::Data {
+        vbuf: vertex_buffer,
+        out: main_color,
+});
+{% endhighlight %}
+
+We match all the specified compile/run-time information (names, formats, offsets, field etc) with the shader reflection upon creating a PSO. This allows catching the possible mismatches at the initialization time, leaving the draw time error-less. We also provide a range of PSO components to be used in that definition, including some trade-offs between compile time and run-time safety. For example, you can have your RTV format known in advance, or you can pass it in as data during the PSO creation. Please refer to the dedicated [blog post](http://gfx-rs.github.io/2016/01/22/pso.html) for more information about how PSOs are implemented in GFX.
+
+
+## Conclusion
+
+As you can see, we borrowed a lot of concepts from the new graphics APIs. This allows us to stay as close to the underlying hardware as possible, while still be portable. Dx12 has the closest programming model to GFX, despite the fact that we don't have the corresponding backend in the works. Plus, GFX adds a bit of Rusty flavor to the API, in particular with the bind-less approach.
+
+Hopefully, this post would shed some light on the complexity that GFX brings along, and allow more people to get familiar with it. We strive to be the best graphics abstraction library out there, and we appreciate any help or feedback. Check out our [Readme](https://github.com/gfx-rs/gfx/blob/master/README.md) and visit us on [Gitter](https://gitter.im/gfx-rs/gfx). As the last word, big thanks to all the old and new contributors that made GFX possible!
```

Comment:
I'd rather not dive into static dispatch since it would require to expand on all sources of overhead in GFX. This deserves its own blog post, eventually ;)